beacon-chain/db/kv/state_diff_helpers.go

```diff
@@ -0,0 +1,233 @@
+	return buf
+}
+
+func (s *Store) getAnchorState(offset uint64, lvl int, slot primitives.Slot) (anchor state.ReadOnlyBeaconState, err error) {
+	if lvl <= 0 || lvl > len(flags.Get().StateDiffExponents) {
+		return nil, errors.New("invalid value for level")
+	}
+
+	if uint64(slot) < offset {
+		return nil, ErrSlotBeforeOffset
+	}
+	relSlot := uint64(slot) - offset
+	prevExp := flags.Get().StateDiffExponents[lvl-1]
+	span := math.PowerOf2(uint64(prevExp))
+	anchorSlot := primitives.Slot(uint64(slot) - relSlot%span)
+
+	// anchorLvl can be [0, lvl-1]
+	anchorLvl := computeLevel(offset, anchorSlot)
+	if anchorLvl == -1 {
+		return nil, errors.New("could not compute anchor level")
+	}
+
+	// Check if we have the anchor in cache.
+	anchor = s.stateDiffCache.getAnchor(anchorLvl)
+	if anchor != nil {
+		return anchor, nil
+	}
+
+	// If not, load it from the database.
+	anchor, err = s.stateByDiff(context.Background(), anchorSlot)
+	if err != nil {
+		return nil, err
+	}
+
+	// Save it in the cache.
+	err = s.stateDiffCache.setAnchor(anchorLvl, anchor)
+	if err != nil {
+		return nil, err
+	}
+	return anchor, nil
+}
+
+// computeLevel computes the level in the diff tree. Returns -1 in case slot should not be in tree.
+func computeLevel(offset uint64, slot primitives.Slot) int {
+	rel := uint64(slot) - offset
+	for i, exp := range flags.Get().StateDiffExponents {
+		span := math.PowerOf2(uint64(exp))
+		if rel%span == 0 {
+			return i
+		}
+	}
+	// If rel isn’t on any of the boundaries, we should ignore saving it.
+	return -1
+}
+
+func (s *Store) setOffset(slot primitives.Slot) error {
+	err := s.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(stateDiffBucket)
+		if bucket == nil {
+			return bbolt.ErrBucketNotFound
+		}
+
+		offsetBytes := bucket.Get(offsetKey)
+		if offsetBytes != nil {
+			return fmt.Errorf("offset already set to %d", binary.LittleEndian.Uint64(offsetBytes))
+		}
+
+		offsetBytes = make([]byte, 8)
+		binary.LittleEndian.PutUint64(offsetBytes, uint64(slot))
+		if err := bucket.Put(offsetKey, offsetBytes); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Save the offset in the cache.
+	s.stateDiffCache.setOffset(uint64(slot))
+	return nil
+}
+
+func (s *Store) getOffset() uint64 {
+	return s.stateDiffCache.getOffset()
+}
+
+func keyForSnapshot(v int) []byte {
+	switch v {
+	case version.Fulu:
+		return fuluKey
+	case version.Electra:
+		return ElectraKey
+	case version.Deneb:
+		return denebKey
+	case version.Capella:
+		return capellaKey
+	case version.Bellatrix:
+		return bellatrixKey
+	case version.Altair:
+		return altairKey
+	default:
+		// Phase0
+		return []byte{}
+	}
+}
+
+func addKey(v int, bytes []byte) ([]byte, error) {
+	key := keyForSnapshot(v)
+	enc := make([]byte, len(key)+len(bytes))
+	copy(enc, key)
+	copy(enc[len(key):], bytes)
+	return enc, nil
+}
+
+func (s *Store) decodeStateSnapshot(enc []byte) (state.BeaconState, error) {
```

Comment:
```suggestion
func decodeStateSnapshot(enc []byte) (state.BeaconState, error) {
```

This doesn't need to be a pointer receiver.